### PR TITLE
Migrate from ethers v5 to v6

### PR DESCRIPTION
**Pull request by Stanback**
### Description:

This pull request addresses the migration from ethers.js v5 to v6. The changes include updates to dependencies, code adjustments, and modifications to the demos to ensure compatibility with the new version.

For more information, see [Migrating from v5](https://docs.ethers.org/v6/migrating/)

### Changes:

* Update package.json to use ethers v6
* Modify the codebase to utilize the updated ethers API
  * One thing to note is that it now extends AbstractProvider instead of JsonRpcProvider, this is because some of the properties (e.g. nextId) are now private, and it didn't seem necessary or beneficial to extend the RPC provider
* Update the demos to match the new ethers API and fix any breaking changes
* Update the documentation to reflect the new ethers version and API changes

### Testing:

Both demos have been updated. Additionally, manual testing was conducted to ensure that it's functioning as expected with no regressions.

Please review the changes and provide feedback or suggestions if necessary. It's in need of a more thorough review, testing and discussion before it should be merged. It might be best to keep this in a parallel branch / release for awhile. I'm happy to make any additional updates or modifications to ensure a smooth migration.

Thank you for considering this pull request.

### CODE REVIEW

File: package.json
Comment (raster21) [inline package.json:34]: shouldn't the peerDependency on `ethers` also be upgraded to 6.x?

Comment (Stanback) [reply package.json:34]: @raster21 Good callout, I've updated the peer dependency to match 👍

**Review by Rubilmax — APPROVED** [review]

**Review by raster21 — APPROVED** [review]

**Comment by EdoAPP** [comment]
Any updates on this? Why hasn't it been merged?

**Comment by alexx855** [comment]
image

i want to just play around with flashbots and i can't, my project its already using ethers v6

File: src/index.ts
Comment (Ranguna) [inline src/index.ts:1012]: If the `raw` property no longer exists, should these comments be removed for simplicity ?

File: src/index.ts
Comment (Ranguna) [inline src/index.ts:595]: What if `nonce[address]` is `0n` ?

Should we use `??` instead of `||` here ?

File: src/index.ts
Comment (Ranguna) [inline src/index.ts:577]: Should we use `bigint` here instead of `number` ?

**Comment by Ranguna** [comment]
Seems like new changes have been added, which include usages of ethers v5 `BigNumber` class, see here for more info: https://github.com/Stanback/ethers-provider-flashbots-bundle/compare/feature/ethers6...flashbots:ethers-provider-flashbots-bundle:master

Although I do wonder how productive it is to keep patching new changes if there are no maintainers around to merge this PR sooner or later :(

We might as well just publish the changes in this PR to npm and abandon the original package.

**Comment by Stanback** [comment]
Thanks for the code review @Ranguna! 🙏  I'll go ahead and make those changes and publish a v6 version to npm for folks to use in the interim.

**Comment by Ranguna** [comment]
Man, you are a lifesaver!

Although I ended up moving to the mev share client for now, I'm sure there are many other people whole appreciate the v6 version of this package!

**Comment by Ranguna** [comment]
Thanks for the effort @Stanback !

So far so good, although I haven't tested any frontrun strategies. The examples from mev share only include backrunning (which makes sense, since it's the whole point of these private mempools).

But will see if I can test frontruns soon.

**Comment by gholam-deblan** [comment]
Thanks for preparing the migration that is much needed but given the lack of reaction from the owners, I will not even bother testing this library and will look into mev-share-client directly, for front-running though.

**Comment by ape-dev-cs** [comment]
Is there anything that's currently blocking this PR from being merged?

**Comment by slavivanov** [comment]
I'm not a contributor to the library, but reviewed the code and it looks good to me. Hopefully, someone from the maintainers can review it.

**Comment by BlankerL** [comment]
Seems @Stanback did a bunch of amazing commits. Code looks pretty neat to me, I think they should be merged and Ethers v6 is inevitable.

**Comment by AmazingAng** [comment]
Let it Merge

**Comment by metachris** [comment]
Quick update -- this PR is on our radar and a v6 update is being prepared now.

**Comment by AmazingAng** [comment]
Thx!

**Comment by Nenad7955** [comment]
what is the ETA on this merge?

**Comment by pre-vail** [comment]
_i would have posted this as an issue in [flashbots-ethers-v6-provider-bundle](https://github.com/Stanback/ethers-provider-flashbots-bundle) but it looks like issues have not been enabled_

i get an error when running [src/demo.ts](https://github.com/Stanback/ethers-provider-flashbots-bundle/blob/master/src/demo.ts). first, i had to make a few changes to the script in order to make it work with ethers v6 but they involved migrating from BigNumber to BigInt so i know my changes are not the reason for the error

when i test the demo on my local (foundry) anvil fork using an alchemy rpc i get this:
```
userStats { error: { message: 'rpc method is not whitelisted', code: -32001 } }
userStatsV2 { error: { message: 'rpc method is not whitelisted', code: -32001 } }
Simulation Error: rpc method is not whitelisted
```

when i test the demo on sepolia i get:
```
userStats {
  error: {
    message: 'the method flashbots_getUserStats does not exist/is not available',
    code: -32601
  }
}
userStatsV2 {
  error: {
    message: 'the method flashbots_getUserStatsV2 does not exist/is not available',
    code: -32601
  }
}
Simulation Error: err: %!w(<nil>); txhash 0x2b59945369fae8e0a8dc412f7534db001cc97a6ea55a19fa43d5c695b4f4554e
```

fyi, i'm console logging the `userStats` object at [line 43](https://github.com/Stanback/ethers-provider-flashbots-bundle/blob/aa5e48ee9b9610ed778ba477bb981490f5ce1769/src/demo.ts#L43) and i have `process.env.TEST_V2` set to `true`, thus the logging of both `userStats` and `userStatsV2`

what am i doing wrong?

**Comment by epheph** [comment]
@pre-vail you are likely using the "protect" url (rpc.flashbots.net) instead of the flashbots relay url (relay.flashbots.net)

**Comment by epheph** [comment]
lgtm, i'll merge it. Question though, what do you think of separating it by major version number, so
v0.x.x - ethers5
v1.0.0 - ethers6
Seems like a reasonable way to separate, and we can maintain the v0 branch as well?

**Comment by AmazingAng** [comment]
Yes,

> lgtm, i'll merge it. Question though, what do you think of separating it by major version number, so v0.x.x - ethers5 v1.0.0 - ethers6 Seems like a reasonable way to separate, and we can maintain the v0 branch as well?

lgtm

**Pull request merged**
